### DISCUSSION TITLE
Transforming a node causes it to lose external props

### DISCUSSION THREAD

**Discussion by pklavins** [josdejong/mathjs#2726] [Ideas]
I have an expression tree, in which some nodes have been given extra properties (such as `node.depth = 1` or `node.priority = 4`). However, it seems that putting that expression tree through the `transform()` function removes those props entirely:

```
    nodeTree = nodeTree.transform((node) => {
        if (node.priority == priority && node.depth == depth) {
          changed = true;
          return new math.expression.node.ConstantNode(node.eval());
        } else return node;
      });
```
I even tried embedding the data within the `node.comment` property, but that gets erased too. I did check the value of `node` in the `else return` section and it did contain my added `depth` and `priority` props, so something in the transform callback is deleting them.

**Comment by josdejong** [comment]
Thanks for bringing this up. When transforming a tree, the nodes of the tree are cloned and you will indeed lose any added properties. I think we could either:

1. Introduce a special property `attributes` (an object) which can be used to put custom data alongside a node, which is neatly copied when cloning the node.
2. When cloning a node, blindly copy all foreign properties that are encountered on the object at hand. 

I think I prefer the first solution. 

Anyone interested in implementing this?
